### ISSUE TITLE
Current state of DenyHosts: maintenance, upstream packages, failing tests

### ISSUE THREAD

**Issue by petems** [denyhosts/denyhosts#70]
Hi All,

So, I've been playing with Denyhosts, and I find it way more intuitive and easy to setup than fail2ban 😄 

* Travis tests failing: Right now the travis tests have been failing on commits for about 2 years: https://travis-ci.org/denyhosts/denyhosts/builds

* Original sync host unmaintained: http://xmlrpc.denyhosts.net/ seems to be unmaintained, and hasn't had new reports since 2011. I noticed the work @janpascal was doing over at https://github.com/janpascal/denyhosts_sync, would it be a good idea to have a "blessed" upstream host to use and mention it in the man page?

* Age of upstream packages: Most people who use denyhosts are probably using the upstream packages from CentOS, Ubuntu etc. Most of them are quite out of date and. I noticed a few of the people in the issues page are maintainers of the upstream packages (I think @jasontibbitts is the Fedora package maintainer?)

As an avid denyhosts user for a few years who's only recently looked into what it's current state is, I'd be happy to help with any issues. I'm not python expert but I can do what I can to help, and have plenty of infrastructure and automation knowledge to configure systems if needed 🎆

**Comment by jasontibbitts** [comment]
As a package maintainer, I have gotten a bit behind because I've never really found the time to make the integration with the systemd journal work properly.  The branch that's horrendously out of date is EPEL6, which I never wanted branched in the first place and don't actually want to maintain.  The others are merely a bit behind, and I even did a little work on them recently that I need to push out.

As for the sync server, the one at denyhosts.net is that old closed-source one run by the original Denyhosts guy.  It's not really connected to the Denyhosts project these days as I understand things.  My instance is still running at dhsync.math.uh.edu but I guess everyone who was using it stopped.  I don't know what happened to getting some more generic hostname for it, since I think the current project maintainers own denyhost.com.  I should probably try updating it to 2.2.0 as I recall that was supposed to have some code for allowing multiple sync servers to participate in a network.  I'm not sure that there are any others besides mine, though.

**Comment by adriel** [comment]
denyhost.com domain is for sale...  Seems there hasn't been any denyhosts releases since the end of 2015, and that was a pre-release, seems like this is a dead/dying project :(

**Comment by janpascal** [comment]
Hi all - the project isn't dead yet, but it's not looking too good either. The sync server Jason was running stopped functioning after a library update. @jasontibbitts, can I help you getting it back on line? Or else, I have Debian packages (and the knowledge to fix them if need be) for the sync server, maybe @petems has a Debian server (since you seems to offer help in infrastructure) that could be used, at least for a while, to run a canonical sync server? As for myself, I would like to put some effort into the code quality of denyhosts so that we can get a proper release done in a couple of weeks/months...

**Comment by adriel** [comment]
What kinda server do you need, spec wise? I may be able to help.

**Comment by janpascal** [comment]
That would be great! Initially, nothing special is needed, just a server that has a reliable 24/7 internet connection and something like 256M (or maybe 512M) memory. If the new sync server will be used by a serious number (millions) of denyhosts installations it may need more, but we can solve that when the need arrives.

**Comment by adriel** [comment]
Any particular location in mind for the server?
These are my options https://azure.microsoft.com/en-us/regions/

The lowest spec Azure VM is the Basic A0 (1 core, 0.75 GB memory), so that should cover your needs for now.

I've got free monthly credit at Azure so doesn't cost me anything.
When it starts to get used allot it'll need to be migrated off to somewhere else as it'll hit my free usage limit.

Let me know if you want to move forward on this and I'll create the VM. 
I'll need your SSH public key so you can log into it, I'll also need your email so you can open ports etc on Azure. Though I'm happy to do that if you're not familiar with Azure.

**Comment by janpascal** [comment]
I've no experience with docker images. The sync server is apt-gettable. Add the line
`deb http://www.vanbest.org/reprepro/ unstable main contrib non-free`
to your /etc/apt/sources.list. The key used to sign the repository is at `http://www.vanbest.org/janpascal/debian-archive-key.asc`. That may make is easy to turn it into a Docker image, if not, I would be more than happy to use an Azure VM if you can provide one. I don't care much where it is located, and I'll send you my SSH public key separately.

**Comment by adriel** [comment]
Just looking into creating a docker image for this. I added the repo and the key, but then doing an `apt update` I get invalid sig errors.

   add-apt-repository "deb http://www.vanbest.org/reprepro/ unstable main contrib non-free"
   wget -qO - http://www.vanbest.org/janpascal/debian-archive-key.asc | apt-key add -

    # apt update
    Ign:1 http://ftp.nz.debian.org/debian stretch InRelease
    Hit:2 http://ftp.nz.debian.org/debian stretch-updates InRelease                                       
    Hit:3 http://ftp.nz.debian.org/debian stretch Release                                                                                                 
    Hit:5 http://security.debian.org/debian-security stretch/updates InRelease                                                                            
    Get:6 http://www.vanbest.org/reprepro unstable InRelease [6,840 B]                                    
    Err:6 http://www.vanbest.org/reprepro unstable InRelease
      The following signatures were invalid: DD03DDF48E51495C820CB356729D9E9563217C63
    Reading package lists... Done
    W: GPG error: http://www.vanbest.org/reprepro unstable InRelease: The following signatures were invalid: DD03DDF48E51495C820CB356729D9E9563217C63
    E: The repository 'http://www.vanbest.org/reprepro unstable InRelease' is not signed.
    N: Updating from such a repository can't be done securely, and is therefore disabled by default.
    N: See apt-secure(8) manpage for repository creation and user configuration details.

**Comment by jasontibbitts** [comment]
So if you folks are going to do this, can I just drop the existing setup and not worry about trying to get it to work?  The only problem I have is the fact that it relies on some javascript minimizer which isn't packaged for Fedora and which would take far more effort for me to use than the sync server itself.  If it weren't for that my instance would have been fixed months ago.

**Comment by janpascal** [comment]
I've seen the same kind of thing after updating to stretch... Could you please check I've given you the right repository key, by checking that apt-key list shows the key that apt-get update is complaining about? 

Adriel Kloppenburg <notifications@github.com> schreef op 7 juli 2017 12:10:21 CEST:
>Just looking into creating a docker image for this. I added the repo
>and the key, but then doing an `apt update` I get invalid sig errors.
>
>add-apt-repository "deb http://www.vanbest.org/reprepro/ unstable main
>contrib non-free"
>wget -qO - http://www.vanbest.org/janpascal/debian-archive-key.asc |
>apt-key add -
>
>    # apt update
>    Ign:1 http://ftp.nz.debian.org/debian stretch InRelease
>Hit:2 http://ftp.nz.debian.org/debian stretch-updates InRelease        
>                              
>Hit:3 http://ftp.nz.debian.org/debian stretch Release                  
>                                                                       
>Hit:5 http://security.debian.org/debian-security stretch/updates
>InRelease                                                              
>             
>Get:6 http://www.vanbest.org/reprepro unstable InRelease [6,840 B]     
>                              
>    Err:6 http://www.vanbest.org/reprepro unstable InRelease
>The following signatures were invalid:
>DD03DDF48E51495C820CB356729D9E9563217C63
>    Reading package lists... Done
>W: GPG error: http://www.vanbest.org/reprepro unstable InRelease: The
>following signatures were invalid:
>DD03DDF48E51495C820CB356729D9E9563217C63
>E: The repository 'http://www.vanbest.org/reprepro unstable InRelease'
>is not signed.
>N: Updating from such a repository can't be done securely, and is
>therefore disabled by default.
>N: See apt-secure(8) manpage for repository creation and user
>configuration details.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/denyhosts/denyhosts/issues/70#issuecomment-313642749

-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by adriel** [comment]
Hmm, it looks to be the same.

      nic@debian:~$ sudo apt update
      Hit:1 http://security.debian.org/debian-security stretch/updates InRelease
      Ign:2 http://ftp.nz.debian.org/debian stretch InRelease        
      Hit:3 http://ftp.nz.debian.org/debian stretch-updates InRelease
      Hit:4 http://ftp.nz.debian.org/debian stretch Release
      Get:6 http://www.vanbest.org/reprepro unstable InRelease [6,840 B]
      Err:6 http://www.vanbest.org/reprepro unstable InRelease
        The following signatures were invalid: DD03DDF48E51495C820CB356729D9E9563217C63
      Reading package lists... Done
      W: GPG error: http://www.vanbest.org/reprepro unstable InRelease: The following signatures were invalid: DD03DDF48E51495C820CB356729D9E9563217C63
      E: The repository 'http://www.vanbest.org/reprepro unstable InRelease' is not signed.
      N: Updating from such a repository can't be done securely, and is therefore disabled by default.
      N: See apt-secure(8) manpage for repository creation and user configuration details.
      nic@debian:~$ sudo apt-key list
      /etc/apt/trusted.gpg
      --------------------
      pub   dsa1024 2007-05-11 [SC]
            DD03 DDF4 8E51 495C 820C  B356 729D 9E95 6321 7C63
      uid           [ unknown] Jan-Pascal van Best (Debian archive key) <janpascal@vanbest.org>
      sub   elg2048 2007-05-11 [E]

**Comment by adriel** [comment]
Ok, so I went ahead and tried installing the server package (ignoring the cert errors for now), but when I run apt-get install;

    apt-get --allow-unauthenticated --no-install-recommends install denyhosts-server

I get can't connect to MySQL errors:

    ERROR 2002 (HY000): Can't connect to local MySQL server through socket 
    '/var/run/mysqld/mysqld.sock' (2 "No such file or directory").
    unable to connect to mysql server.

Am I meant to setup a working MySQL server first?
It looked like the `denyhosts-server` package was doing that, since it asked me to set MySQL passwords etc.

Full log here: https://pastebin.com/ytceEhbU

PS, also tested this on a clean install of Debian in VMware, with the same MySQL errors.

**Comment by janpascal** [comment]
Yes, it needs a working MySQL server, and I believe that's why I made it a Recommends... When making Docker images, I don't know what you should do about this kind of dependency. On a normal Debian server the Recommends should take  care of installing MySQL first so that the post-installation script of denyhosts-server can take care of creating the database and  tables on it. 

Adriel Kloppenburg <notifications@github.com> schreef op 8 juli 2017 05:32:05 CEST:
>Ok, so I went ahead and tried installing the server package (ignoring
>the cert errors for now), but when I run apt-get install;
>
>apt-get --allow-unauthenticated --no-install-recommends install
>denyhosts-server
>
>I get can't connect to MySQL errors:
>
>ERROR 2002 (HY000): Can't connect to local MySQL server through socket 
>    '/var/run/mysqld/mysqld.sock' (2 "No such file or directory").
>    unable to connect to mysql server.
>
>Am I meant to setup a working MySQL server first?
>It looked like the `denyhosts-server` package was doing that, since it
>asked me to set MySQL passwords etc.
>
>Full log here: https://pastebin.com/ytceEhbU
>
>PS, also tested this on a clean install of Debian in VMware, with the
>same MySQL errors.
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/denyhosts/denyhosts/issues/70#issuecomment-313830591

-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by adriel** [comment]
Yeah I saw that later one, was half asleep. With Docker DB's are usualy separate from the application, so I'm going to try do that, then manually add the DB settings to the denyhosts server.

**Comment by adriel** [comment]
Well I think I got it working: http://one.australiaeast.cloudapp.azure.com:9911/
Still needs a bit of work behind the scenes, but mostly done I'd say.

**Comment by janpascal** [comment]
Great work, I'll have my ssh server sync to your new sync server tonight. By the way, I just registered denyhosts.org, I'll have sync.denyhosts.org be a cname record for the azure server you just fixed. 

Cheers

Jan-Pascal

**Comment by petems** [comment]
Awesome work @adriel and @janpascal! I registered `denyhosts.party` a few months ago, have been meaning to get it up and working, then we can add it to the pool also 😄

**Comment by janpascal** [comment]
I've created a CNAME record for sync.denyhosts.org for the server you've 
installed. My denyhosts client is successfully reporting IPs using this 
address. The statistics page doesn't seem to reflect this, though. Could 
you have a look in the denyhosts-server log file?

**Comment by adriel** [comment]
Hmm, this is all I'm getting in the logs: https://pastebin.com/aTmYt3xV

Also I see I'm on v2.2.0, is the v2.2.2 not been packaged yet?

**Comment by adriel** [comment]
So I've commented out line 433 in stats.py `yield Registry.DBPOOL.runInteraction(make_country_bargraph)` which has gotten the web frontend to update somewhat correctly now. But with no bargraphs etc.

I'm not very familiar with Python and not to sure what to do to fix the 'NoneType' error.

**Comment by janpascal** [comment]
I've seen issues earlier generating the stats when the database is (almost) empty, it looks about the same. I'll look into it somewhere today (CEST) 

Adriel Kloppenburg <notifications@github.com> schreef op 9 juli 2017 12:29:36 CEST:
>So I've commented out line 433 in stats.py `yield
>Registry.DBPOOL.runInteraction(make_country_bargraph)` which has gotten
>the web frontend to update somewhat correctly now. But with no
>bargraphs etc.
>
>I'm not very familiar with Python and not to sure what to do to fix the
>'NoneType' error.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/denyhosts/denyhosts/issues/70#issuecomment-313911682

-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by janpascal** [comment]
Yeah, something's fishy until there is at least some report in the database, and then until a full day later because some statistics are only calcatulated once per day. Let's see how it looks like tomorrow.

Maybe some of the other people reading this could are sync.denyhosts.org as their sync server in their denyhosts configuration?

**Comment by janpascal** [comment]
Just uploaded the 2.2.3-1 Debian package to my repository. It should fix the current problems. I've also update the key used to sign the repository from a 1024 bit DSA to a 4096 bit RSA key, which seem to fix the error messages `apt-key update`  gave.

**Comment by adriel** [comment]
Nice! I've updated the docker image and it does indeed seem to have done the trick.

I've been trying to setup denyhosts (client) in docker, and it seems to be going well, just one question, what is the purpose of `--sync`? 
Is it needed to sync with the azure server (when turned on in the config)?
e.g. `python /usr/local/bin/denyhosts.py --config /etc/denyhosts.conf --sync --foreground`

It's not talked about on the denyhosts_sync GitHub readme page in the setup steps for denyhosts (client)

**Comment by adriel** [comment]
Just checked the logs for denyhosts-server and found 1 error (repeating):

    /usr/share/denyhosts-server/denyhosts_server/stats.py:186: RankWarning: Polyfit may be poorly conditioned
      z = numpy.polyfit(x_num, y, 1)

Not sure if it's important or not as it seems to be working.

**Comment by adriel** [comment]
Finally finished the Docker image, it's [here](https://hub.docker.com/r/adriel/denyhosts-sync/) if anyone is interested.
It's up and running on my server now too, with a daily backup of the DB, though haven't decided how/where to store the database yet (only locally at the moment).

**Comment by adriel** [comment]
**Important, DNS changes needed within 3 weeks.**

The free Azure server I've been using to host the DenyHosts Sync server for the last few years is coming to an end (out of my control), but I have some good news.
I've moved the database to a different server which seems to handle the load better, so now I only need to host the front end, which uses much less resources, so I've spun up the low spec "forever" free vm from google which now hosts the front end and connects to the off site database.

This new server is hosted in the US, (vs Australia, where the old one was), which should result in lower latency for most people.
I've already closed down everything on the Azure server, it's just proxying everything to the new server till it expires.

The old Azure server will **close down on the 24th, June**, at which point I'll lose the `one.australiaeast.cloudapp.azure.com` sub-domain.

@janpascal when you have second could you update the cname for `sync.denyhosts.org` to point to `denyhosts.adriels.com`?

I think only the sync.denyhosts.org needs updating, since it [appears](https://mxtoolbox.com/SuperTool.aspx?action=a%3axmlrpc.denyhosts.net&run=toolpage) xmlrpc.denyhosts.net cname is sync.denyhosts.org.
